# MeRIPseqPipe

**MeRIP-seq analysis pipeline arranged multiple alignment tools, peakCalling tools, Merge Peaks\' methods and methylation analysis methods.**

[![Nextflow](https://img.shields.io/badge/nextflow-%E2%89%A519.04.0-brightgreen.svg)](https://www.nextflow.io/)
[![check in Biotreasury](https://img.shields.io/badge/Biotreasury-collected-brightgreen)](https://biotreasury.rjmart.cn/#/tool?id=61140)
[![install with bioconda](https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg)](http://bioconda.github.io/)
[![Docker](https://img.shields.io/docker/automated/kingzhuky/meripseqpipe.svg)](https://hub.docker.com/r/kingzhuky/meripseqpipe)

## Introduction

Here, we present MeRIPseqPipe, an integrated analysis pipeline for MeRIP-seq data based on Nextflow. It integrates ten main functional modules including data preprocessing, quality control, read mapping, peak calling, peak merging, motif searching, peak annotation, differential methylation analysis, differential expression analysis, and data visualization, which covers the basic analysis of MeRIP-seq data. 

All the analysis modules are generated by [Nextflow](https://www.nextflow.io/), and all the third-party tools are encapsulated in the [Docker](https://www.docker.com/resources/what-container) container.

## Quick Start

1. install [`nextflow`](https://nf-co.re/usage/installation)

2. pull [`docker`](https://docs.docker.com/engine/installation/) image from dockerhub: [`kingzhuky/meripseqpipe:dev`](https://hub.docker.com/r/kingzhuky/meripseqpipe)

3. cloning this repository

    ```bash
    git clone https://github.com/canceromics/MeRIPseqPipe.git
    nextflow run /path/to/MeRIPseqPipe --help
    ```

4. test it on a minimal dataset with a single command

    ```bash
    nextflow run path/to/meripseqpipe -profile test,docker
    ```

5. Start running your own analysis!

    ```bash
    nextflow run path/to/meripseqpipe -profile docker --designfile designfile.tsv --comparefile compare.txt -resume --aligners star --fasta hg38_genome.fa --gtf gencode.v25.annotation.gtf --rRNA_fasta hg38_rRNA.fasta --outdir path/to/results --skip_createbedgraph --peakMerged_mode rank --star_index hg38/starindex --skip_meyer --skip_matk --methylation_analysis_mode Wilcox-test
    ```

See [usage docs](docs/usage.md) for more details and all of the available options when running the pipeline.

## Documentation

The MeRIPseqPipe documentation is split into the following files:

1. [Usage](docs/usage.md)
    * [Parameter Documentation](docs/parameter_docs.md)
    * An overview of how the pipeline works, how to run it and a description of all of the different command-line flags.
    * let us know if you need more customization!
2. [Output](docs/output.md)
    * An overview of the different results produced by the pipeline

## Pipeline overview

This pipeline is built using Nextflow and integrates tools as follows:

* **Quality control and preprocessing of raw data**
  * [fastp](https://github.com/OpenGene/fastp): quality trimming and adapter clipping
  * [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/): generate quality reports
  * [RSeQC](http://rseqc.sourceforge.net/): assess mapping performance to give more insight into data quality
* **Read alignment**
  * [STAR](https://github.com/alexdobin/STAR): Spliced Transcripts Alignment to a Reference
  * [HISAT2](http://daehwankimlab.github.io/hisat2/): memory efficient splice aware alignment to a reference
  * [TopHat2](https://ccb.jhu.edu/software/tophat/index.shtml): accurate alignment of transcriptomes in the presence of insertions, deletions and gene fusions
  * [BWA](https://github.com/lh3/bwa): fast and accurate short read alignment with Burrows-Wheeler transform
* **Peak calling**
  * [MACS2](https://github.com/macs3-project/MACS): Model-based Analysis of ChIP-seq
  * [MeTPeak](https://github.com/compgenomics/MeTPeak): a novel, graphical model-based peak-calling method
  * [MATK](http://matk.renlab.org/): a deep learning-based MeRIP-seq analysis tool at single-nucleotide-resolution
  * [Meyer](https://pubmed.ncbi.nlm.nih.gov/22608085/): a peak-calling tool based on Fisher's exact test
* **Peak merging**
  * [RobustRankAggreg](https://pubmed.ncbi.nlm.nih.gov/22247279/): a rank aggregation algorithm
  * [MSPC](https://genometric.github.io/MSPC/): using combined evidence from replicates to evaluate ChIP-seq peaks
  * [BEDTools](https://bedtools.readthedocs.io/en/latest/): using "mergeBed" and "intersectBed" function
* **Peak annotation**
  * Perl scripts: peak start/end position, gene start/end position, transcript ID, strand, gene type (coding or noncoding, lncRNA or mRNA, etc.), peak location, gene ensemble ID, etc.
  * [annotatePeaks.pl](http://homer.ucsd.edu/homer/ngs/annotation.html): whether a peak is in the TSS (transcription start site), TTS (transcription termination site), Exon (Coding), 5' UTR Exon, 3' UTR Exon, Intronic, or Intergenic and also shows the distance to TSS
* **Motif searching**
  * [HMOER](http://homer.ucsd.edu/homer/motif/): Hypergeometric Optimization of Motif EnRichment
* **M6A sites predicition**
  * [MATK](http://matk.renlab.org/): predict m6A sites at single nucleotide resolution
* **Differential expression analysis**
  * [featureCounts](https://pubmed.ncbi.nlm.nih.gov/24227677/): read counting relative to gene biotype
  * [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html): for differential expression analysis of RNA-Seq, SAGE-Seq, ChIP-Seq or HiC count data
  * [edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html): for differential expression analysis of RNA-Seq, SAGE-Seq, ChIP-Seq or HiC count data
* **Differential methylation analysis**
  * [QNB](https://pubmed.ncbi.nlm.nih.gov/28859631/): a statistical approach for differential RNA methylation analysis with count-based small-sample sequencing data
  * [MATK](http://matk.renlab.org/): using a Bayesian hierarchical model to eliminate the effect of basal expression and quantify the true m6A level by Markov Chain Monte Carlo sampling
  * Wilcox-test: results are generated by custom R scripts referred to RPKM methods
  * [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html): use a generalized linear model to detect changes in IP coverage while controlling for differences in Input coverage
  * [edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html): use a generalized linear model to detect changes in IP coverage while controlling for differences in Input coverage
* **Report**
  * [MultiQC](https://multiqc.info/): summarize all results from quality control and alignment
  * R packages

## Credits

MeRIPseqPipe was originally written by Xiaoqiong Bao, Kaiyu Zhu.

## Contributions and Support

If you would like to contribute to this pipeline, please see the [contributing guidelines](.github/CONTRIBUTING.md).

## BioTreasury

MeRIPseqPipe has been registed to BioTreasury(https://biotreasury.rjmart.cn/#/tool?id=61140), welcome to use and comment!

## Citation
Xiaoqiong Bao, Kaiyu Zhu, Xuefei Liu, Zhihang Chen, Ziwei Luo, Qi Zhao, Jian Ren, Zhixiang Zuo, MeRIPseqPipe: an integrated analysis pipeline for MeRIP-seq data based on Nextflow, Bioinformatics, 2022;, btac025, https://doi.org/10.1093/bioinformatics/btac025.

## Acknowledgements
Thanks to nf-core for the support and guidance!

You can cite the nf-core publication as follows:

> The nf-core framework for community-curated bioinformatics pipelines.
> 
> Philip Ewels, Alexander Peltzer, Sven Fillinger, Harshil Patel, Johannes Alneberg, Andreas Wilm, Maxime Ulysse Garcia, Paolo Di Tommaso & Sven Nahnsen.
> Nat Biotechnol. 2020 Feb 13. doi: 10.1038/s41587-020-0439-x.
